include/fixed_containers/recursive_reflection.hpp

```diff
@@ -0,0 +1,373 @@
+#include "fixed_containers/type_name.hpp"
+
+#include <bitset>
+#include <variant>
+
+// This library aims to provide a reflection strategy for recursive reflection into a type.
+// We provide default reflection strategy for primitives, optional, iterable and reflectable types.
+// User can also extend the reflection strategy by providing a custom reflection strategy for a
+// type. We even allow overriding default reflection strategy for a type by providing a
+// STRATEGY_NO_DEFAULT specialization and a custom reflection strategy.
+
+// constants
+namespace fixed_containers::recursive_reflection_detail
+{
+inline constexpr std::string_view ITERABLE_PATH_NAME = "data[:]";
+inline constexpr std::string_view OPTIONAL_PATH_NAME = "value()";
+inline constexpr std::string_view PATH_DELIMITER = ".";
+}  // namespace fixed_containers::recursive_reflection_detail
+
+// reflection related concepts for dfs strategy and metadata exrtraction strategy in struct_view
+namespace fixed_containers::recursive_reflection_detail
+{
+using reflection::ReflectionConstructible;
+
+template <typename T>
+concept DurationLike = requires {
+    typename T::rep;
+    typename T::period;
+    typename std::is_arithmetic<typename T::rep>::type;
+    { T::min() } -> std::same_as<T>;
+    { T::max() } -> std::same_as<T>;
+    { T::zero() } -> std::same_as<T>;
+};
+
+template <typename T>
+concept ResizableIterable = requires {
+    typename T::value_type;
+    { T{}.size() } -> std::same_as<std::size_t>;
+    { T{}.capacity() } -> std::same_as<std::size_t>;
+    { T{}.resize(std::declval<std::size_t>()) } -> std::same_as<void>;
+};
+
+template <typename T>
+inline constexpr bool IS_OPTIONAL_IMPL = false;
+
+template <typename T>
+inline constexpr bool IS_OPTIONAL_IMPL<std::optional<T>> = true;
+
+template <typename T>
+concept IsOptional = IS_OPTIONAL_IMPL<T>;
+
+template <typename T>
+concept Bitset = requires { requires std::same_as<T, std::bitset<T{}.size()>>; };
+
+// details of supported primitives
+template <typename T>
+concept EnumValue = std::is_enum_v<T>;
+
+template <typename T>
+concept EnumView = std::same_as<T, std::string_view>;
+
+// The list of supported primitives
+template <typename T>
+concept PrimitiveValue = std::is_arithmetic_v<T> || DurationLike<T> || Bitset<T>;
+
+template <typename T>
+concept PrimitiveView = std::is_pointer_v<T> && PrimitiveValue<std::remove_pointer_t<T>>;
+
+template <typename T>
+concept Primitive = PrimitiveValue<T> || PrimitiveView<T>;
+
+// exclude some contiguous sized ranges that we do not want to consider as Iterable
+// use inline constexpr bool for customizability
+template <typename T>
+inline constexpr bool NOT_CONSIDERED_ITERABLE = false;
+
+template <>
+inline constexpr bool NOT_CONSIDERED_ITERABLE<std::string_view> = true;
+
+template <IsOptional T>
+inline constexpr bool NOT_CONSIDERED_ITERABLE<T> = true;
+
+template <Bitset T>
+inline constexpr bool NOT_CONSIDERED_ITERABLE<T> = true;
```

Comment:
`std::bitset` and `std::string` are special-cases because we don't want to recurse into them, so maybe rename to express that?

Reply:
`NOT_CONSIDERED_ITERABLE` indeed means not considered iterable.